Revive python linting by file

ddn0 opened the pull request:
The logic on how pylint searches for pyproject.toml and how it treats
directories versus file is subtle. Instead, just pass pylint files and
its configuration directly.

arthurp commented:
@gurbinder533, you are the "primary review" in my mind. @ddn0, I just want a check on my CI, build, and dependency changes.

@insertinterestingnamehere, you may also be interested in this PR. Feel free to add yourself as a reviewer if you want to be fully involved.

insertinterestingnamehere commented:
It's super-exciting that this is working. I have limited availability over the next few days, but I can hopefully look through it later next week. Just don't let my review hold anything up.

arthurp commented on python/galois/_loops.pxd.jinja:
```diff
@@ -0,0 +1,47 @@
+# distutils: language=c++
+# distutils: extra_compile_args=["-std=c++17"]
+
+# THIS FILE IS GENERATED.
```
Hahahaha. I could hide the "GENERATED" warning so it's not visible in the template. I hadn't bothered, but I would actually prefer it.

gurbinder533 commented on python/galois/_bfs_property_graph.pyx:
```diff
@@ -1,10 +1,35 @@
+cdef extern from * nogil:
+    """
+std::shared_ptr<arrow::Table> MakeTable(const std::string& name,
+                                        const std::vector<uint32_t>& data) {
```
We won't need this function when we move to python/numa and use pyarrow, right?

arthurp replied:
This is the only remaining direct access to Arrow C++. I left this in because the pyarrow wrappers are missing proper arrow builders. The python interface doesn't use builders and instead relies on the user building a data structure as either a Python list or as a numpy data structure and then converting it to an arrow data structure (This might be zero-copy, BTW, I'm not sure). This code could be replaced with numpy usage.

gurbinder533 commented on python/galois/numba/galois.py:
```diff
@@ -0,0 +1,58 @@
+
+UserContextType = SimpleNumbaPointerWrapper(galois._loops.UserContext)
+
+UserContextType.register_method("push", ctypes.CFUNCTYPE(None, ctypes.c_voidp, ctypes.c_uint64))
```
Is this type (c_uint64) specifying what type can be pushed? In other words, do we need to register_methods for different types?

arthurp replied:
Responses are in the main conversation on the PR.

gurbinder533 commented on tests/test_property_graph.py:
```diff
@@ -0,0 +1,115 @@
+    prop = g.get_node_property("length")
+    out = np.empty((g.num_nodes(),), dtype=int)
+
+    do_all(0, len(g), func_operator(g, prop, out), "operator")
```
This is great. 
Can we also pass galois native data structures such as galois::LargeArrays and galois::gstl::Vector to the python do_all operators? 
Also, we should do some performance benchmarking for the do_all and for_each just to make sure.

arthurp replied:
Responses are in the main conversation on the PR.

arthurp commented:
Re: the iteration types of galois loops and contexts

`uint64` is the ONLY type currently supported by galois-python loops. My thinking was that all node and edge ids are uint64 and that's a good time for indicies into almost anything, so it's an acceptable value type (the type passed to the loop body callback) for most cases.

However, I had already realized I need to support InsertBag or something similar (I was trying to implement BFS). And types like InsertBag don't support indexing. So there is clearly a need for support for multiple iterator types (It may be OK to stick to uint64_t as the value type). Supporting multiple iterator types will be possible, but tricky to do with good performance. I need to think through exactly how to abstract iterators at the python level in a way that they can be unwrapped effectively when passed to C++.

Allowing multiple value types would be harder since it would require instancing the loops (do_all, for_each) and contexts. It should be doable, but it will probably require instancing things for every value type we support, so the value types will not be extensible.

gurbinder533 commented on tests/test_loops.py:
```diff
@@ -0,0 +1,72 @@
+    total = 0
+    def f(i):
+        nonlocal total
+        total += i
```
f is run by do_all. Just for my understanding, is "total" incremented atomically in parallel by multiple threads?

gurbinder533 replied:
Or is this a single thread execution as it doesn't have a "do_all_operator" decorator?

arthurp replied:
Execution is in multiple threads, but this is a normal python function, so it takes the GIL before executing so calls to the function are sequentialized. So it's not really parallel, but there are multiple threads. This is never something people would actually use to get good performance, but it's useful for debugging and development potentially.

ddn0 committed:
Revive python linting by file

The logic on how pylint searches for pyproject.toml and how it treats
directories versus file is subtle. Instead, just pass pylint files and
its configuration directly.

ddn0 committed:
Tweak Python linter rules

arthurp commented on pyproject.toml:
```diff
@@ -98,7 +101,6 @@ enable = [
   "bare-except",
   "binary-op-exception",
   "boolean-datetime",
-  "broad-except",
```
I would like to keep this enabled and disable it on a line by line basis. The reason is that an overly broad except can hide `TypeError`, `KeyboardInterrupt`, and in some cases even `SyntaxError`. How many places would need to explicitly disable this warning?
